### PR TITLE
Slower Online -> target

### PR DESCRIPTION
**Pull request by sandhawalia**
1. Slower online to target 
2. Warm start 
3. Data gen script to see how severe the augmentation is

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Can you do the pull request against https://github.com/daniel-j-h/visiont-pytorch/pull/1 the `vit-next` branch, because master is quite outdated and if you do it against master github shows us all changes and not just your changes vs my branch.

File: visiont/scripts/data_gen.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    RandomCrop,
+    RandomHorizontalFlip,
+    RandomApply,
+    RandomGrayscale,
+    ColorJitter,
+    GaussianBlur,
+    RandomRotation,
+    RandomErasing,
+)
+from skimage.io import imsave
```
Comment (daniel-j-h) [inline visiont/scripts/data_gen.py:21]: Why not PIL, we have it already

File: visiont/scripts/data_gen.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+from visiont.scripts.train import ImageDirectory, Convert
+
+
+def save_samples(image1, image2, idx, path):
+    def save_tensor_to_image(fpath, img):
+        img = img.squeeze(0)
+        arr = img.cpu().numpy()
+        arr = (np.transpose(arr, (1, 2, 0)) * 255).astype(np.uint8)
+        imsave(fpath.as_posix(), arr)
```
Comment (daniel-j-h) [inline visiont/scripts/data_gen.py]: Why not

    torchvision.transforms.ToPILImage(mode=None)

Comment (daniel-j-h) [reply visiont/scripts/data_gen.py]: Also we have einops, say no to raw transpose etc :hugs:

File: visiont/scripts/data_gen.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    save_tensor_to_image(path.joinpath(f"{idx:06}-1.jpg"), image1)
+    save_tensor_to_image(path.joinpath(f"{idx:06}-2.jpg"), image2)
```
Comment (daniel-j-h) [inline visiont/scripts/data_gen.py]: this is a good idea for visualization, but we can we not just dump images in the train script when we have them at hand anyway?

File: visiont/scripts/data_gen.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+def main(args):
+
+    # Just a list to try out different augmentation ideas here
+    # Bring learnings into visiont.script.train.main
+    transform = Compose(
+        [
+            Convert("RGB"),
+            Resize(1024),
+            RandomCrop(640),
```
Comment (daniel-j-h) [inline visiont/scripts/data_gen.py]: This has to be in line with how we do it in train, no?

I think a better idea is to have a helper function which writes input1, input2 tensors in the train loop for debugging purpose. If we use this separate generate tool, then it's easy to have them out of sync and dump one thing but we train on another thing.

File: visiont/scripts/train.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            RandomApply(
+                [ColorJitter(brightness=0.8, contrast=0.8, saturation=0.8, hue=0.2)],
+                p=0.5,
+            ),
+            RandomApply([GaussianBlur((3, 3), (1.5, 1.5))], p=0.5),
+            RandomGrayscale(p=0.5),
+            RandomRotation(degrees=(-10, 10)),
+            ToTensor(),
+            Normalize(mean=mean, std=std),
+            RandomErasing(p=0.5),
```
Comment (daniel-j-h) [inline visiont/scripts/train.py:129]: Why after totensor and normalize?

File: visiont/scripts/train.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            optimizer.step()
+
+            # After training the online network, we transfer
+            # a weighted average of the weights to the target
+            # Only apply warmup criterion on epoc == 0
+            # Update target network so if "lags" the online network
+            if idx > warmup_steps and epoch == 0:
+                update_target()
+            elif idx % target_update_steps == 0 and epoch > 0:
+                update_target()
```
Comment (daniel-j-h) [inline visiont/scripts/train.py]: Why do we work with steps, but then special case the first epoch? Can warmup not work on steps alone?

The step var counts the total ("global") steps anyway.

**Comment by daniel-j-h** [comment]
Have you read the BYOL and SimCL papers and are replicating their augmentations or is this just trying things out?

Also do you have resources on the slower online to target and warm start thingie? Was it not working for you without?

**Comment by sandhawalia** [comment]
Lets get it working first we can make cosmetic changes later :) All are good change requests

File: visiont/scripts/train.py (hunk 6 above)
Comment (sandhawalia) [reply visiont/scripts/train.py]: idx is local to the epoch. Warmup is only applied to first epoch

File: visiont/scripts/train.py (hunk 5 above)
Comment (sandhawalia) [reply visiont/scripts/train.py:129]: From pytorch source examples : https://pytorch.org/docs/stable/_modules/torchvision/transforms/transforms.html#RandomErasing

File: visiont/scripts/data_gen.py (hunk 4 above)
Comment (sandhawalia) [reply visiont/scripts/data_gen.py]: keeping larger size to have a better view of images removing normalization etc which are not required for visalization

Comment (sandhawalia) [reply visiont/scripts/data_gen.py]: > This has to be in line with how we do it in train, no?
> 
> I think a better idea is to have a helper function which writes input1, input2 tensors in the train loop for debugging purpose. If we use this separate generate tool, then it's easy to have them out of sync and dump one thing but we train on another thing.

I think we could publish training samples to tensboardX

File: visiont/scripts/data_gen.py (hunk 3 above)
Comment (sandhawalia) [reply visiont/scripts/data_gen.py]: This is just to find before we do any training to do a sanity check on transforms

File: visiont/scripts/data_gen.py (hunk 2 above)
Comment (sandhawalia) [reply visiont/scripts/data_gen.py]: Da

**Comment by sandhawalia** [comment]
Batch : 128 on Two Titan V. Small training set `46000` images. Previously it used to collapse in 10 epochs. Each epoch is 324 iterations. 

```
[1300] loss : 470.4345358276367
[2800] loss : 448.2505807495117
[3100] loss : 441.7204022216797
[3800] loss : 425.84991943359375
[5000] loss : 387.1744251098633
[9500] loss : 290.0904251098633
[17700] loss : 131.59521484375
[22600] loss : 66.92655303955078
[27800] loss : 25.206183166503905
[32400] loss : 10.247162523269653
```

**Comment by sandhawalia** [comment]
Nevermind still collapsing
